ITR/data/base_providers.py

```diff
@@ -358,44 +388,63 @@ def get_company_intensity_and_production_at_base_year(self, company_ids: List[st
+        cols = df.columns.tolist()        
+        df = df[cols[1:3] + [cols[0]] + cols[3:]]
+        return df
 
     def get_company_fundamentals(self, company_ids: List[str]) -> pd.DataFrame:
         """
         :param company_ids: A list of company IDs
         :return: A pandas DataFrame with company fundamental info per company (company_id is a column)
         """
-        return pd.DataFrame.from_records(
-            [ICompanyData.parse_obj(c.dict()).dict() for c in self.get_company_data(company_ids)],
-            exclude=['projected_targets', 'projected_intensities', 'historic_data']).set_index(
-                self.column_config.COMPANY_ID)
+        excluded_cols = ['projected_targets', 'projected_intensities', 'historic_data']
+        df = pd.DataFrame.from_records(
+            [ICompanyData.parse_obj({k:v for k, v in c.dict().items() if k not in excluded_cols}).dict()
+             for c in self.get_company_data(company_ids)]).set_index(self.column_config.COMPANY_ID)
+        return df
 
-    def get_company_projected_trajectories(self, company_ids: List[str]) -> pd.DataFrame:
+    def get_company_projected_trajectories(self, company_ids: List[str], year=None) -> pd.DataFrame:
         """
         :param company_ids: A list of company IDs
-        :return: A pandas DataFrame with projected intensity trajectories per company, indexed by company_id
-        """
-        trajectory_list = [self._convert_projections_to_series(c, self.column_config.PROJECTED_EI, EScope.S1S2)
-                           for c in self.get_company_data(company_ids)]
-        if trajectory_list:
-            with warnings.catch_warnings():
-                # pd.DataFrame.__init__ (in pandas/core/frame.py) ignores the beautiful dtype information adorning the pd.Series list elements we are providing.  Sad!
-                warnings.simplefilter("ignore")
-                return pd.DataFrame(trajectory_list)
+        :param year: values for a specific year, or all years if None
+        :return: A pandas DataFrame with projected intensity trajectories per company, indexed by company_id and scope
+        """
+        projected_trajectories = [(c.company_id, scope, getattr(c.projected_intensities, scope.name).projections)
+                                  # FIXME: we should make _companies a dict so we can look things up rather than searching every time!
+                                  for c in self._companies for scope in EScope.get_result_scopes()
+                                  if c.company_id in company_ids
+                                  if getattr(c.projected_intensities, scope.name)]
+        if projected_trajectories:
+            if year:
+                company_ids, scopes, values = list(map(list, zip(*[(pt[0], pt[1], [yvp.value for yvp in pt[2] if yvp.year==year][0]) for pt in projected_trajectories])))
+                index=pd.MultiIndex.from_tuples(zip(company_ids, scopes), names=["company_id", "scope"])
```

Comment:
these 2 lines seem duplicated for "if" and "else" branches of the conditional check.
maybe we could optimize, by placing them outside of if-else

Reply:
These are different comprehensions, depending on whether building a pd.Series (list comprehension) or pd.DataFrame (dict comprehension).  Can you write the code you think would be better?

Reply:
i tried, but couldn't
i agree with your approach here